### PR TITLE
Update Custom Components feature to match final RFC.

### PR DESCRIPTION
**Pull request by rwjblue**
* Removes view hierarchy updating (not included in RFC)
* Updates custom manager hook method names to match final RFC
* Add `capabilities` system (still very basic)
* Disentangle curly component manager from custom component manager.
* Add optional capabilities (`destructor` / `asyncLifecycleCallbacks`)
* Add temporary global (private on the global, public via modules API)
  for `capabilities`
* Migrate `setComponentManager` to use `WeakMap`'s instead of mutation

(paired with @chancancode on some of the gnarly typescripting)

---

TODO:

- [x] Add more tests based on the examples from the RFC
- [x] Add tests using angle bracket invocation

Future TODO:

- [ ] Determine impact / APIs needed for proper ember-inspector support

### CODE REVIEW

File: packages/ember-glimmer/lib/component-managers/custom.ts [hunk 1]
```diff
@@ -1,32 +1,119 @@
...
+): ComponentCapabilityMask {
+  assert('Invalid component manager compatibility specified', managerAPI === '3.4');
+
+  return {
+    asyncLifeCycleCallbacks: !!options.asyncLifecycleCallbacks,
+    destructor: !!options.destructor,
+  };
+}
+
+export interface DefinitionState<ComponentInstance = Opaque> {
```
Comment (rwjblue) [inline packages/ember-glimmer/lib/component-managers/custom.ts]: from kris: lets remove the defaulted `Opaque`'s in these interfaces and swap to pass `Opaque` when we actually use them

File: packages/ember-glimmer/lib/resolver.ts [hunk 2]
```diff
@@ -320,27 +322,41 @@ export default class RuntimeResolver implements IRuntimeResolver<OwnedTemplateMe
...
+        );
+
+        let definition = new CustomManagerDefinition(
+          name,
+          component,
+          manager,
+          layout || meta.owner.lookup<OwnedTemplate>(P`template:components/-default`)
+        );
+        this.componentDefinitionCache.set(key, definition);
+        return definition;
```
Comment (rwjblue) [inline packages/ember-glimmer/lib/resolver.ts]: Need to call the `finalizer()` before this `return`

File: packages/ember-glimmer/lib/renderer.ts [hunk 3]
```diff
@@ -32,7 +32,7 @@ export type IBuilder = (env: Environment, cursor: Cursor) => ElementBuilder;
 
 export class DynamicScope implements GlimmerDynamicScope {
   constructor(
-    public view: Component | {} | null,
```
Comment (rwjblue) [inline packages/ember-glimmer/lib/renderer.ts]: Since we removed the dynamic scope manipulation in this PR, we can revert this change

File: packages/ember-glimmer/lib/resolver.ts [hunk 4]
```diff
@@ -353,4 +369,16 @@ export default class RuntimeResolver implements IRuntimeResolver<OwnedTemplateMe
 
     return definition;
   }
+
+  _lookupComponentManager(owner: Owner, managerId: string): CustomComponentManager {
+    if (this.customManagerCache.has(managerId)) {
+      return this.customManagerCache.get(managerId)!;
+    }
+    let delegate = owner.lookup<ManagerDelegate>(`component-manager:${managerId}`);
+    let manager = new CustomComponentManager(delegate);
```
Comment (rwjblue) [inline packages/ember-glimmer/lib/resolver.ts]: Lets also pass the `owner` to the constructor here. It will allow us to more easily participate in the renderer's view registry when we want to make it work for the Ember Inspector.

File: packages/ember-glimmer/lib/component-managers/custom.ts [hunk 5]
```diff
@@ -1,32 +1,119 @@
...
+  name: string;
+  ComponentClass: Factory<ComponentInstance>;
+  symbolTable: ProgramSymbolTable;
+  template?: any;
+}
+
+export interface ComponentCapabilityMask {
```
Comment (chancancode) [inline packages/ember-glimmer/lib/component-managers/custom.ts]: Probably just call this `Capabilities`. We probably want to mark the "content" of this private somehow, since from the user's/public API perspective, the function is returning an opaque (and explicitly unstable between versions) thing.

File: packages/ember-glimmer/lib/component-managers/custom.ts [hunk 6]
```diff
@@ -1,32 +1,119 @@
...
+  asyncLifeCycleCallbacks: boolean;
+  destructor: boolean;
+}
+
+export interface CustomComponentManagerArgs {
```
Comment (chancancode) [inline packages/ember-glimmer/lib/component-managers/custom.ts]: TODO move into glimmer

File: packages/ember-glimmer/lib/component-managers/custom.ts [hunk 7]
```diff
@@ -1,32 +1,119 @@
...
+  extends ManagerDelegate<ComponentInstance> {
+  destroyComponent(instance: ComponentInstance): void;
+}
+
+export interface ComponentArguments {
+  positional: Opaque[];
+  named: Dict<Opaque>;
+}
+
+function valueForCapturedArgs(args: CapturedArguments): CustomComponentManagerArgs {
```
Comment (chancancode) [inline packages/ember-glimmer/lib/component-managers/custom.ts]: TODO move into glimmer

File: packages/ember-glimmer/lib/component-managers/custom.ts [hunk 8]
```diff
@@ -52,125 +139,129 @@ export interface ComponentArguments<T = {}> {
...
     super();
   }
 
   create(
     _env: Environment,
     definition: DefinitionState,
     args: Arguments,
-    dynamicScope: DynamicScope
-  ): CustomComponentState<T> {
+    _dynamicScope: DynamicScope
```
Comment (chancancode) [inline packages/ember-glimmer/lib/component-managers/custom.ts]: if we don't need it we can just remove the arg right?

(where did we land w.r.t. view tracking again?)

File: packages/ember-glimmer/lib/component-managers/custom.ts [hunk 9]
```diff
@@ -52,125 +139,129 @@ export interface ComponentArguments<T = {}> {
...
+export default class CustomComponentManager<ComponentInstance = Opaque>
+  extends AbstractComponentManager<CustomComponentState<ComponentInstance>, DefinitionState>
+  implements
+    WithStaticLayout<
+      CustomComponentState<ComponentInstance>,
+      DefinitionState<ComponentInstance>,
+      OwnedTemplateMeta,
+      RuntimeResolver
+    > {
+  constructor(private delegate: ManagerDelegate<ComponentInstance>) {
```
Comment (chancancode) [inline packages/ember-glimmer/lib/component-managers/custom.ts]: I am not actually sure if we need this to be a stateful object. The "definition state" and "custom component state" already has the "delegate" in it so we don't have to store it again on `this`. Then this would just be stateless and can be stored in a constant. It's one fewer allocation per component, not sure if it matters much.

File: packages/ember-glimmer/lib/component-managers/custom.ts [hunk 10]
```diff
@@ -52,125 +139,129 @@ export interface ComponentArguments<T = {}> {
...
+    let invocationArgs = valueForCapturedArgs(capturedArgs);
+    const component = delegate.createComponent(definition.ComponentClass, invocationArgs);
 
     return new CustomComponentState(delegate, component, capturedArgs);
   }
 
-  update({ component, args }: CustomComponentState<T>) {
-    this.delegate.update(component, args.value());
+  update({ component, args }: CustomComponentState<ComponentInstance>) {
+    this.delegate.updateComponent(component, valueForCapturedArgs(args));
```
Comment (chancancode) [inline packages/ember-glimmer/lib/component-managers/custom.ts]: so, for example this would be

```ts
update({ delegate, component, args }: CustomComponentState<ComponentInstance>) {
  delegate.updateComponent(component, valueForCapturedArgs(args));
}
```

File: packages/ember-glimmer/lib/utils/custom-component-manager.ts [hunk 11]
```diff
@@ -1,45 +1,27 @@
...
+const getPrototypeOf = Object.getPrototypeOf;
+const MANAGERS: WeakMap<any, string> = new WeakMap();
 
-export function componentManager(obj: any, managerId: String) {
-  if ('reopenClass' in obj) {
-    return obj.reopenClass({
-      [COMPONENT_MANAGER]: managerId,
-    });
-  }
+export function setComponentManager(managerId: string, obj: any) {
```
Comment (chancancode) [inline packages/ember-glimmer/lib/utils/custom-component-manager.ts]: The `set` language in the RFC supposedly give us the license to mutate the class. Would we want to do that instead of a manual prototype chain walk?

**Review by chancancode — APPROVED** [review]

File: packages/ember-glimmer/lib/utils/custom-component-manager.ts (hunk 11 above)
Comment (rwjblue) [reply packages/ember-glimmer/lib/utils/custom-component-manager.ts]: The prototype walk is actually quite fast, and if we move this to setting on the object we break any objects that `Object.seal` or `Object.freeze` themselves (which is actually a good practice, and somewhat common).

Imagine this implementation for "template only components":

```js
owner.register('component-manager:pseudo-template-only', EmberObject.extend({
  capabilities: capabilities('3.4'),

  createComponent() {
    return null;
  },

  updateComponent() {},

  getContext() {
    return null;
  },
});

const BASE_COMPONENT = Object.freeze({});

setComponentManager('pseudo-template-only, BASE_COMPONENT);
```

If we were to mutate the object, we loose this...

File: packages/ember-glimmer/lib/component-managers/custom.ts (hunk 8 above)
Comment (rwjblue) [reply packages/ember-glimmer/lib/component-managers/custom.ts]: Confirm, will remove.

We landed on: `TODO: Make this work with the Ember Inspector` 😝

File: packages/ember-glimmer/lib/utils/custom-component-manager.ts (hunk 11 above)
Comment (chancancode) [reply packages/ember-glimmer/lib/utils/custom-component-manager.ts]: ah yes I remember discussing this

File: packages/ember-glimmer/lib/component-managers/custom.ts (hunk 7 above)
Comment (rwjblue) [reply packages/ember-glimmer/lib/component-managers/custom.ts]: https://github.com/glimmerjs/glimmer-vm/pull/832 implements the interface in glimmer-vm. I'll follow up with a PR to remove the duplication once that lands.
